_overviews/scala3-book/domain-modeling-fp.md

```diff
@@ -0,0 +1,408 @@
+---
+title: FP Modeling
+type: section
+description: This chapter provides an introduction to FP domain modeling with Scala 3.
+num: 22
+previous-page: domain-modeling-oop
+next-page: methods-intro
+---
+
+
+This chapter provides an introduction to functional programming (FP) style domain modeling in Scala 3. When modeling the world around us with FP, you typically use these Scala constructs:
+
+- Enumerations
+- Case classes
+- Traits
+
+
+## Introduction
+
+In FP, the *data* and the *operators on that data* are two separate things; you aren’t forced to encapsulate them together like you do with OOP.
+
+The concept is like numerical algebra. When you think about whole numbers whose values are greater than or equal to zero, you have a *set* of possible values that looks like this:
+
+````
+0, 1, 2 ... Int.MaxInt
+````
+
+Ignoring the division of whole numbers, the possible *operators* on those values are:
+
+````
++, -, *
+````
+
+An FP design is implemented in a similar way:
+
+- You have a set of values
+- You have a collection of operators that work on those values
+
+In this chapter we’ll model the data and operations for a “pizza” in a pizza store. You’ll see how to implement the “data” portion of the Scala/FP model, and then you’ll see several different ways you can organize the operations on that data.
+
+
+
+## Modeling the data
+
+Modeling the “data” portion of a domain model in Scala is simple:
+
+- Create ADTs and GADTs with the `enum` construct
+- Use `case` classes for more complicated systems
+<!-- TODO: wording on that last line? -->
+
+
+### Crust size, crust type, and toppings
+
+Data concepts like crust size, crust type, and toppings are concisely modeled with the Scala 3 `enum` construct:
+
+```scala
+enum CrustSize:
+  case Small, Medium, Large
+
+enum CrustType:
+  case Thin, Thick, Regular
+
+enum Topping:
+  case Cheese, Pepperoni, BlackOlives, GreenOlives, Onions
+```
+
+
+### Modeling a pizza
+
+Now you can model a `Pizza` class using a `case` class:
+
+```scala
+import CrustSize._
+import CrustType._
+import Topping._
+
+case class Pizza(
+  crustSize: CrustSize,
+  crustType: CrustType,
+  toppings: Seq[Topping]
+)
+```
+
+And that’s it. That’s the data model for an FP-style pizza system. This solution is very concise because it doesn’t require the operations on a pizza to be combined with the data model. The data model is easy to read, like declaring the design for a relational database.
+
+<!-- TODO: compare this with C-style structs and others? -->
+
+#### More of the data model
+
+Because that was very short, you may prefer to see more of the data model for an entire pizza-ordering system. Here are a few other `case` classes that are used to model such a system:
```

Comment:
```suggestion
Because that was very short, you might consider this data modeling approach for the entire pizza-ordering system. Here are a few other `case` classes that are used to model such a system:
```